File: examples/LearningHadoop/MapReduceRecSysJava/src/main/java/com/example/UserVectorToCooccurenceReduce.java
package com.example;


import org.apache.hadoop.mapreduce.Reducer;
import org.apache.mahout.math.RandomAccessSparseVector;
import org.apache.mahout.math.VarIntWritable;
import org.apache.mahout.math.Vector;
import org.apache.mahout.math.VectorWritable;

import java.io.IOException;

/**
 * Created by andryw on 20/02/15.
 */
public class UserVectorToCooccurenceReduce extends Reducer<VarIntWritable,VarIntWritable,VarIntWritable,VectorWritable> {

    public void reduce(VarIntWritable itemIndex1, Iterable<VarIntWritable> itemIndex2s, Context context) throws IOException, InterruptedException {
        Vector cooccureenceRow = new RandomAccessSparseVector(Integer.MAX_VALUE,100);
        for (VarIntWritable indexWritable2 : itemIndex2s){
            int index2 = indexWritable2.get();
            cooccureenceRow.set(index2,cooccureenceRow.get(index2)+1);
        }
        context.write(itemIndex1,new VectorWritable(cooccureenceRow));
    }
}

File: nMiners/src/nMiners/API/UserVectorToCooccurenceReduce.scala
package API

import java.util.Iterator
import Utils._

import org.apache.hadoop.mapred.{Reporter, OutputCollector, Reducer, MapReduceBase}
import org.apache.mahout.math.{VarIntWritable, RandomAccessSparseVector, VectorWritable, VarLongWritable}

/**
 * Created by arthur on 06/04/15.
 */
class UserVectorToCooccurenceReduce extends MapReduceBase with Reducer[VarIntWritable,VarIntWritable,VarIntWritable,VectorWritable]{

  override def reduce(itemIndex1: VarIntWritable, itemIndex2s: Iterator[VarIntWritable], outputCollector: OutputCollector[VarIntWritable, VectorWritable], reporter: Reporter) = {

    var cooccureenceRow = new RandomAccessSparseVector(Integer MAX_VALUE, 100);
    println(itemIndex2s.isEmpty)


    itemIndex2s.foreach((item: VarIntWritable) => {
      val itemIndex2 = item.get();

      val oldValue = cooccureenceRow get (itemIndex2)
      cooccureenceRow set(itemIndex2, oldValue + 1)

    })

    outputCollector.collect(itemIndex1,new VectorWritable(cooccureenceRow))
  }

}
